### PR TITLE
Feature ocs redesign

### PR DESCRIPTION
**Pull request by Nthemba**
This PR has changes to the ODataConnectedService. The aim is to remove clutter from the ConfigODataEndpoint page and fit all elements to the page without scrolling.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/OData/ODataConnectedService?pullRequest=249) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/OData/ODataConnectedService?pullRequest=249) before we can accept your contribution.<br/><br/>:x: Nthemba [sign now](https://cla.dotnetfoundation.org/OData/ODataConnectedService?pullRequest=249)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/OData/ODataConnectedService?pullRequest=249) it.</sub>

**Comment by KenitoInc** [comment]
@Nthemba Share some screenshot (before and after)

**Review by habbes — CHANGES_REQUESTED** [review]
There's a bug introduced with these changes: the proxy configuration options are not applied to the web request when fetching the metadata.

The issue is [on this line of `ConfigODataEndpointViewModel`](https://github.com/OData/ODataConnectedService/blob/d30f286dbd40d24eb69a3485248bce47699c0caa/src/ODataConnectedService/ViewModels/ConfigODataEndpointViewModel.cs#L102).

The `UserSettings.IncludeWebProxy` property is false, it was bound to the "Include WebProxy" checkbox, but that has been removed. Maybe you can remove that property altogether and test whether `UserSettings.WebProxyHost` is null or empty in the `if` statement. You may need to make similar changes to the T4 template as well. Alternatively you can set `UserSettings.IncludeWebProxy` to true when `UserSettings.WebProxyHost` is neither null nor empty.

**Comment by habbes** [comment]
It appears there are a few failing tests, not sure if these are a result of this PR's changes or a regression introduce in previous PRs:

image

cc @ElizabethOkerio

**Comment by habbes** [comment]
Here are some before and after screenshots cc @KenitoInc 

**Before**

OCSCurrentCustomConfig

**After**

OCSNewWebConfigTab
OCSNewWebConfigTabWithCredentials

OCSNewCustomHeadersTab

**Comment by habbes** [comment]
@Nthemba I don't think the **Include Custom Config** checkbox is necessary since the tab buttons already trigger visibility of the config fields. It just adds an unnecessary extra step in my opinion.

File: src/ODataConnectedService/Views/ConfigODataEndpoint.xaml [hunk 1]
```diff
@@ -84,136 +84,184 @@
...
-                    TextWrapping="Wrap" />
+                <ToggleButton
+                    x:Name="WebProxyButton"
+                     Grid.Column="1"
+                    Width="220"
+                    Height="20"
+                    Margin="0,5,20,5"
+                    HorizontalAlignment="Right"
+                    IsEnabled="True"
+                    IsChecked="{Binding UserSettings.WebProxyButton, Mode=TwoWay}" Checked="WebProxyButton_Checked" Grid.ColumnSpan="2" >
```
[the remark this answers is not in the document]
Comment (Nthemba) [reply src/ODataConnectedService/Views/ConfigODataEndpoint.xaml]: No there wasn't, I factored the property name to use the existing code.

**Comment by habbes** [comment]
@Nthemba I've noticed a certain issue, the **Include Custom Config** is not really applied once you've edited the custom headers or proxy fields. Here's an example:
- Add a new connected service
- enable **Include Custom Config** checkbox
- Edit the custom headers, then delete the contents (this will result in `UserSettings.CustomHeaders` being `""` as opposed to `null`)
- Edit the proxy host and delete the contents (this will result in `UserSettings.WebProxy` being `""`
- Uncheck the **Include Custom Config**
- Click "Next"

This will result in errors due to the headers (or proxy) being invalid. The wizard tried to apply the headers and proxy even though **Include Custom Config** was unchecked. It doesn't seem like this checkbox is being used properly (maybe that's another argument for removing it altogether).

I think you should update the condition for headers and proxies to check for nulls or empty and not just for nulls, e.g.:

```c#
if (string.IsNullOrEmpty(UserSettings.CustomHeaders)
{
   ...
}
```

**Comment by gathogojr** [comment]
@Nthemba Here's my feedback/observation covering the first wizard page
One thing I'd like us to address is the inconsistent capitalization of words in the labels, alignment and sizing

Button (Load Settings Config file) => "Load settings config file" or "Load Settings Config File"

TextBox (Service name:) => "Service name" or "Service Name" - to include or exclude colon (REF: "Address" is not followed by colon)

TextBox (Service name:) & Button (Load Settings Config file) => Allow both or just the TextBox to have variable with such that they are only separated by their margins - like the TextBox and TextBlock for "Address"

**Maybe add a new wizard page "Endpoint Settings" to house custom HTTP headers and network credentials given that majority of customers never need to tinker with them?**

CheckBox (Include Custom Http Headers) => "Include custom HTTP headers" or "Include Custom HTTP Headers". Is the alignment okay?

TextBox (Custom Headers:) => "Custom headers" or "Custom Headers" - to include or exclude colon? Is the alignment okay?

TextBlock (for custom headers) => Neater alignment and sizing

CheckBox (Save custom http headers in ConnectedService.json => What is ConnectedService.json to a user? "Save custom HTTP headers to settings config file" or "Save Custom HTTP Headers to Settings Config File". Neater alignment?

CheckBox (Include WebProxy) => "Include web proxy" or "Include Web Proxy". Maybe replace "Include" with "Configure" or "Set" or "Supply" or better word? Maybe "<better word> web proxy details"?

TextBox (Enter the webproxy host with the port (e.g. http://localhost:8080):) => "Enter the web proxy host and port (e.g. http://localhost:8080):"

**Maybe make every TextBox and the related TextBlock be closely packed together same way "Enter the webproxy host with the port (e.g. http://localhost:8080):" and the related TextBlock are**

CheckBox (Include Proxy Network credentials) => Do we really need this? If Username, Password and Domain are not always applicable, we could get rid of this checkbox and adjust the labels of Username, Password and Domain to read "Username (if applicable):", "Password (if applicable):" and "Domain (if applicable):" respectively

CheckBox (Include Proxy Network credentials) => [If we decide to keep it] "Include proxy network credentials" or "Include Proxy Network Credentials". Maybe use "<better word> web proxy network credentials"?

TextBox (Username:)/TextBox => Neater alignment and sizing
TextBox (Password:)/TextBox => Neater alignment and sizing
TextBox (Domain:)/TextBox => Neater alignment and sizing

CheckBox (Save web proxy credentials in ConnectedService.json) => What is ConnectedService.json to a user? "Save web proxy network credentials to settings config file" or "Save Web Proxy Network Credentials to Settings Config File". Neater alignment


**Make improvements to other wizard pages, "Settings" wizard page specifically.**

**Comment by Nthemba** [comment]
image

**Comment by ElizabethOkerio** [comment]
> image

nit: Maybe we can also align the service name label and the field and also the address for uniformity with the other lables and fields.

File: src/ODataConnectedService/ViewModels/ConfigODataEndpointViewModel.cs [hunk 2]
```diff
@@ -88,7 +88,7 @@ internal string GetMetadata(out Version edmxVersion)
             if (!metadataUri.IsFile)
             {
                 var webRequest = (HttpWebRequest)WebRequest.Create(metadataUri);
```
Comment (ElizabethOkerio) [inline src/ODataConnectedService/ViewModels/ConfigODataEndpointViewModel.cs:90]: Can we update the vars to use concrete classes. I know these are not your changes. Am not sure why we agreed to use var in this file. @habbes

**Comment by ElizabethOkerio** [comment]
> @Nthemba I've noticed a certain issue, the **Include Custom Config** is not really applied once you've edited the custom headers or proxy fields. Here's an example:
> 
> * Add a new connected service
> * enable **Include Custom Config** checkbox
> * Edit the custom headers, then delete the contents (this will result in `UserSettings.CustomHeaders` being `""` as opposed to `null`)
> * Edit the proxy host and delete the contents (this will result in `UserSettings.WebProxy` being `""`
> * Uncheck the **Include Custom Config**
> * Click "Next"
> 
> This will result in errors due to the headers (or proxy) being invalid. The wizard tried to apply the headers and proxy even though **Include Custom Config** was unchecked. It doesn't seem like this checkbox is being used properly (maybe that's another argument for removing it altogether).
> 
> I think you should update the condition for headers and proxies to check for nulls or empty and not just for nulls, e.g.:
> 
> ```cs
> if (string.IsNullOrEmpty(UserSettings.CustomHeaders)
> {
>    ...
> }
> ```

There could be a related issue here but I don't think it is related to this PR's changes. If you check Include Web Proxy and type something in the web proxy host field then delete whatever that you had typed, and leave the Include Web Proxy checkbox checked, you get an error when you try to generate code. There could be some missing checks when generating code: 

ScreenShot_4_21_2022_9_01_31_AM

ScreenShot_4_21_2022_9_02_40_AM

**Comment by ElizabethOkerio** [comment]
Am going to approve these changes then @Nthemba you can log any issues that have been raised and that are not related to UI changes.

**Review by ElizabethOkerio — APPROVED** [review]

**Review by KenitoInc — APPROVED** [review]

**Comment by gathogojr** [comment]
@Nthemba This is a much more improved look. A few nits...
- I think "Enter the webproxy host with the port..." should be "Web proxy host with the port..." - No reason for "Enter " since no other TextBlock has it
- "Include Proxy Network Credentials" => "Include Web Proxy Network Credentials"
- I wonder whether "Http" should "HTTP"?
- "In Config Settings File" => "in Config Settings File" (just like in "Code of Conduct", "Terms of Reference", linking words are usually uncapitalized)

**Review by habbes — APPROVED** [review]

**Comment by Nthemba** [comment]
image

**Pull request merged**